docs/advanced/checkpointing.md

```diff
@@ -0,0 +1,108 @@
+
+We call this process a “checkpointing”.
+
+The goal of checkpointing is to ensure that applications can recover after failures and reprocess records from Kafka producing the same results as if the failure never happened.
+
+Currently, Quix Streams provides *At-Least-Once* processing guarantees, which means that each incoming message will be processed, but it may happen multiple times and generate duplicated outputs.
+
+## Under the Hood
+
+- When `Application` starts processing, it initializes the `Checkpoint` object with a fixed lifetime interval in seconds according to the `commit_interval` setting (every 5 seconds by default).
+- The `Checkpoint` object is responsible for keeping track of processed Kafka offsets and pending state transactions.
+- After the message is successfully processed, its offset is marked as processed in the current checkpoint.
+- When checkpoint commits, it will:
+    1. *Produce changelog messages for every pending state update to the changelog topics.*
+
+        Changelog topics are used to back up state stores in case of failures, and they contain records for every updated state key.
+        
+    2. *Flush the Kafka Producer and verify the delivery of every outgoing message both to output and changelog topics.*
+        
+        If some messages are not delivered, the checkpoint will fail.
+        
+        In this case, the messages need to be reprocessed again starting from the previous committed offset.
+        
+    3. *Synchronously commit the topic offsets to Kafka.*
+        
+        If some offsets fail to commit, the checkpoint will fail as well.
+        
+    4. *Flush the pending state transactions to the durable state stores.*
+        
+        If some state stores fail to update, the checkpoint will fail too.
+        
+- After the checkpoint is fully committed, a new one is created and the processing continues.
+- Besides the regular intervals, the checkpoint is also committed when Kafka partitions are rebalanced.
+
+## Recovering the State Stores
+
+In Quix Streams, all state stores are backed up using the changelog topics in Kafka.
+
+Since version 2.5, all changelog messages contain the corresponding offsets of input topic messages.
+
+During recovery, these offsets are compared with the currently committed offset of the input topic.
+
+If the processed offset in the changelog record is higher than the current committed offset (i.e., the update was produced to the changelog, but the Kafka offsets failed to commit), this
+changelog update will not be applied to the state.
+
+This way, all stateful operations will work with a consistent snapshot of the state and produce the same outputs in case of reprocessing.
+
+For more information about changelog topics, see the [**How Changelog Topics Work**](stateful-processing.md#how-changelog-topics-work) section.
+
+## Example Failure Scenarios
```

Comment:
I feel like "examples" would be a bit more...specific or nuanced in nature than what you're outlining here.

To me these are the broad (and most common) failure states that you might expect or reasonably expect to run into (and specifically, ones that we've considered and tried to cover).

That is to say, I'd probably instead call it "Common" or "Typical" Failure Scenarios, instead...with the similar caveat that it's not an exhaustive list.